quantum/keycode_string.c

```diff
@@ -0,0 +1,404 @@
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "keycode_string.h"
+
+#include <string.h>
+#include "bitwise.h"
+#include "keycode.h"
+#include "progmem.h"
+#include "quantum_keycodes.h"
+
+typedef int_fast8_t index_t;
+
+// clang-format off
+/** Names of some common keycodes. */
+static const keycode_string_name_t keycode_names[] = {
+    KEYCODE_STRING_NAME(KC_ENT),
+    KEYCODE_STRING_NAME(KC_ESC),
+    KEYCODE_STRING_NAME(KC_BSPC),
+    KEYCODE_STRING_NAME(KC_TAB),
+    KEYCODE_STRING_NAME(KC_SPC),
+    KEYCODE_STRING_NAME(KC_MINS),
+    KEYCODE_STRING_NAME(KC_EQL),
+    KEYCODE_STRING_NAME(KC_LBRC),
+    KEYCODE_STRING_NAME(KC_RBRC),
+    KEYCODE_STRING_NAME(KC_BSLS),
+    KEYCODE_STRING_NAME(KC_SCLN),
+    KEYCODE_STRING_NAME(KC_QUOT),
+    KEYCODE_STRING_NAME(KC_GRV),
+    KEYCODE_STRING_NAME(KC_COMM),
+    KEYCODE_STRING_NAME(KC_DOT),
+    KEYCODE_STRING_NAME(KC_SLSH),
+    KEYCODE_STRING_NAME(KC_INS),
+    KEYCODE_STRING_NAME(KC_HOME),
+    KEYCODE_STRING_NAME(KC_PGUP),
+    KEYCODE_STRING_NAME(KC_DEL),
+    KEYCODE_STRING_NAME(KC_END),
+    KEYCODE_STRING_NAME(KC_PGDN),
+    KEYCODE_STRING_NAME(KC_RGHT),
+    KEYCODE_STRING_NAME(KC_LEFT),
+    KEYCODE_STRING_NAME(KC_DOWN),
+    KEYCODE_STRING_NAME(KC_UP),
+    KEYCODE_STRING_NAME(KC_TRNS),
+#ifdef EXTRAKEY_ENABLE
+    KEYCODE_STRING_NAME(KC_WBAK),
+    KEYCODE_STRING_NAME(KC_WFWD),
+    KEYCODE_STRING_NAME(KC_WREF),
+    KEYCODE_STRING_NAME(KC_MNXT),
+    KEYCODE_STRING_NAME(KC_MPRV),
+    KEYCODE_STRING_NAME(KC_MPLY),
+    KEYCODE_STRING_NAME(KC_VOLU),
+    KEYCODE_STRING_NAME(KC_VOLD),
+#endif  // EXTRAKEY_ENABLE
+#ifdef MOUSEKEY_ENABLE
+    KEYCODE_STRING_NAME(MS_LEFT),
+    KEYCODE_STRING_NAME(MS_RGHT),
+    KEYCODE_STRING_NAME(MS_UP),
+    KEYCODE_STRING_NAME(MS_DOWN),
+    KEYCODE_STRING_NAME(MS_WHLL),
+    KEYCODE_STRING_NAME(MS_WHLR),
+    KEYCODE_STRING_NAME(MS_WHLU),
+    KEYCODE_STRING_NAME(MS_WHLD),
+#endif  // MOUSEKEY_ENABLE
+#ifdef TRI_LAYER_ENABLE
+    KEYCODE_STRING_NAME(TL_LOWR),
+    KEYCODE_STRING_NAME(TL_UPPR),
+#endif // TRI_LAYER_ENABLE
+#ifdef GRAVE_ESC_ENABLE
+    KEYCODE_STRING_NAME(QK_GESC),
+#endif // GRAVE_ESC_ENABLE
+#ifdef CAPS_WORD_ENABLE
+    KEYCODE_STRING_NAME(CW_TOGG),
+#endif // CAPS_WORD_ENABLE
+#ifdef LAYER_LOCK_ENABLE
+    KEYCODE_STRING_NAME(QK_LLCK),
+#endif // LAYER_LOCK_ENABLE
+    KEYCODE_STRING_NAME(DB_TOGG),
+    KEYCODE_STRING_NAMES_END
+};
+// clang-format on
+/** Users can override this to define names of additional keycodes. */
+__attribute__((weak)) const keycode_string_name_t custom_keycode_names[] = {KEYCODE_STRING_NAMES_END};
```

Comment:
Would be nice to have this available as kb and user options, to keep in line.

Reply:
Thanks for the suggestion! I've revised this with separate _user and _kb tables.